<h1 align="center">Python performance exercises</h1>

## Python best practices exercises

### Exercise 1

considering the following function for concatenating list strings with deliter.

In [7]:
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimeter"""
    res = l_strings[0]
    for e in l_strings[1:]:
        res = res + d + e
    return res

In [16]:
var1 = "foo"
var2 = "bar"
var3 = f"{var1}{var2}"
print(var3) 

foobar


- profile the function and identify the bottlenecks.
- improve speed up of the function
*Hint: you may need to look to the string functions in python documentation*

In [26]:
# write your code here
a = ["Hello","World"]
%timeit ft_concatenate(a,' ')

380 ns ± 20.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


- The bottleneck in in the loop 

using the function join we can avoid using a loop

In [23]:
def ft_concatenate_1(l_strings, d):
    """concatenate list of strings into one string separated by delimeter"""
    return d.join(l_strings)

In [24]:
ft_concatenate_1(a," ")

'Hello World'

In [25]:
%timeit ft_concatenate_1(a," ")

206 ns ± 1.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


### Exercise 2

In this exercise you will solve the following problem using two methods bruteforce mehode, and fast method.

**Problem:** You are given a list of n integers, and your task is to calculate the number of distinct values in the list.

**Example**
- Input:
5
2 3 2 2 3

- Output:
3

**Implement the following methodes:**

1. **bruteforce mehode:** create an empty list and start adding items for the given list without adding the previous item add, at the end the result list will contain unique values, print lenght of the list and you are done. 
2. **fast method** think of using Set data structure.

- time the two methods, what do you think?

In [28]:
# bruteforce fast method
def distinct_brut(a):
    l = [a[0]]
    for i in a :
        if i not in l :
            l.append(i)
    return l,len(l)

In [29]:
a = [ 5 ,2 ,3 ,2 ,2 ,3]
distinct_brut(a)

([5, 2, 3], 3)

In [30]:
%timeit distinct_brut(a)

736 ns ± 34.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [31]:
# write fast method
def distinct_fast(a):
    # insert the list to the set
    list_set = set(a)
    # convert the set to the list
    unique_list = (list(list_set))
    return unique_list, len(unique_list)

In [32]:
distinct_fast(a)

([2, 3, 5], 3)

In [33]:
%timeit distinct_fast(a)

638 ns ± 4.74 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Cython exercises

### Exercise 1

1. load the cython extension.

In [1]:
%load_ext Cython

2. Condidering the following polynome function:

In [2]:
def poly(a,b):
    return 10.5 * a + 3 * (b**2)

- Create an equivalent Cython function of `poly` with name `poly_cy` without any cython improvement, just make its cell a cython cell.

In [49]:
%%cython  


def poly_cy(a,b):
     return 10.5 * a + 3 * (b**2)

/home/zineb/.cache/ipython/cython/_cython_magic_c18c0a5228b85821ff9a96ad916c4db2.c:19:10: fatal error: Python.h: No such file or directory
 #include "Python.h"
          ^~~~~~~~~~
compilation terminated.


3. time the performance of Python and Cython version of the function, what is the factor of speed up here between the two verions.

In [7]:
a = 100
b = 5000
%timeit  poly(a,b)

490 ns ± 69.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [8]:
%timeit  poly_cy(a,b)

473 ns ± 22.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


4. Now lets work on another examples using loop.
    - rewrite the same function below fib that calculate fibonacci series using cython, but now try to add type for the variables used inside it, add a prefix `_cy` to your new cython function.

In [43]:
def fib(n):
    if n==0 :
        return 0
    a, b = 0, 1
    for i in range(n):
        a, b = a + b, a

    return a

In [45]:
fib(5)

5

In [53]:
%%cython -a

cdef int fib_cy( int n):
    if n==0 :
        return 0
    cdef int a = 0
    cdef int b = 1
    for i in range(n):
        a, b = a + b, a

    return a

/home/zineb/.cache/ipython/cython/_cython_magic_28d3855b957d1065dc3dc5f9f1d8258e.c:19:10: fatal error: Python.h: No such file or directory
 #include "Python.h"
          ^~~~~~~~~~
compilation terminated.


- time the two function for fibonacci series, with n = 20, what is the factor of speed now, What do you think?

In [50]:
%timeit fib(20)

1.23 µs ± 83.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [ ]:
%timeit fib_cy(20)

5. Recursive functions are functions that call themselves during their execution. Another interesting property of the Fibonacci series is that it can be written as a recursive function. That’s because each item depends on the values of other items (namely item n-1 and item n-2)

- Rewrite the fib function using recursion. Is it faster than the non-recursive version? Does Cythonizing it give even more of an advantage? 

In [57]:
def Fibo(n):
    # Condition d'arrêt
    if n == 0:
        return 0
    elif n == 1 or n == 2:
        return 1
 
    else:
        #récursivité
        return Fibo(n-1) + Fibo(n-2)
 
Fibo(9)

34

In [58]:
%timeit Fibo(20)

2.17 ms ± 21.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


- The recursive version takes more time than the non recursive one.

In [ ]:
#cythonized version 
%%cython -a

cdef int Fibo_cy( int n):
    # Condition d'arrêt
    if n == 0:
        return 0
    elif n == 1 or n == 2:
        return 1
 
    else:
        #récursivité
        return Fibo(n-1) + Fibo(n-2)

In [ ]:
%timeit Fibo_cy(20)

### Exercise 2

- Monte Carlo methods are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. 
- One of the basic examples of getting started with the Monte Carlo algorithm is the estimation of Pi.

**Estimation of Pi**

- The idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. 
- Imagine a circle inside the same domain with same diameter and inscribed into the square. 
- We then calculate the ratio of number points that lied inside the circle and total number of generated points. 
- Refer to the image below:

![demo](../data/MonteCarloPlot.png)

We know that area of the square is 1 unit sq while that of circle is $\pi \ast  (\frac{1}{2})^{2} = \frac{\pi}{4}$. Now for a very large number of generated points,

![demo](../data/MonteCarloCalc.png)


## The Algorithm

1. Initialize cile_points, square_points and interval to 0.
2. Generate random point x.
3. Generate random point y.
4. Calculate d = x*x + y*y.
5. If d <= 1, increment circle_points.
6. Increment square_points.
7. Increment interval.
8. If increment < NO_OF_ITERATIONS, repeat from 2.
9. Calculate pi = 4*(circle_points/square_points).
10. Terminate.

**Your mission:** time the function `monte_carlo_pi`, identify the bottlenecks and create a new version using cython functionality to speed up monte carlo simulation for PI, use 100,000 points and compare the speed up factor between python and cython, considering the following optimizations:
- add type for variables used.
- add type for the function
- use c rand function instead of python rand function.
 
*Hint: you can import function from C libraries using the following approach `from libc.<name of c library> cimport <library function name>`, replace the holders `<>` with the right identities for the current problem*

In [61]:
import random

def simulate_pi(N):
  n = 0
  for i in range(N):
    x = random.uniform(-1,1)
    y = random.uniform(-1,1)
    if (x**2 + y**2) <= 1 :
      n +=1
  return 4*(n/N)

In [66]:
simulate_pi(1000000)

3.140924

In [67]:
%timeit simulate_pi(1000000)

845 ms ± 6.06 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Numba exercises

### Exercise 1

Previously we considered how to approximateby Monte Carlo.

- Use the same idea here, but make the code efficient using Numba.
- Compare speed with and without Numba when the sample size is large.

In [68]:
from numba import njit

In [69]:
@njit
def simulate_pi_numba(N):
  n = 0
  for i in range(N):
    x = random.uniform(-1,1)
    y = random.uniform(-1,1)
    if (x**2 + y**2) <= 1 :
      n +=1
  return 4*(n/N)

In [70]:
%timeit simulate_pi_numba(1000000)

16.6 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


The execution time improved from 845 ms to 16 ms only.

## Pyccel Exercises

### Exercise 1

considering the following algorithm for binomial coefficient implementation in python.
```python
def factorial(n : int) -> int:
    # to be implemented

def binomial_coefficient(n : int, k : int):
    num = factorial(n)
    den = factorial(k) * factorial(n - k)
    return num // den
```
1. complete the factorial function using recursive methode.
2. profile the `binomial_coefficient`.
3. try to improve factorial function by using iterative method.
4. Use pyccel to accelerate the function.
5. compare the benchmark of the two version python vs pyccel.

In [81]:
# facturial  function
def factorial(n) :
    if (n==1 or n==0) :
        return 1
    else :
        return n * factorial(n - 1); 

In [82]:
factorial(4)

24

In [83]:
#Binomial coefficient
def binomial_coef(n : int, k : int):
    num = factorial(n)
    den = factorial(k) * factorial(n - k)
    return num // den
binomial_coef(16,2)

120

In [85]:
%timeit binomial_coef(160,20)

62.5 µs ± 263 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


The factorial function is time consuming

In [95]:
#improving factorial function
def iter_factorial(n : int):
    f = 1
    if n == 0 :
        return 1
    else :
        for i in range(1,n+1):
            f = f*i
    return f

In [96]:
#Binomial coefficient
def binomial_coef_iter(n : int, k : int):
    num = iter_factorial(n)
    den = iter_factorial(k) * iter_factorial(n - k)
    return num // den
binomial_coef_iter(16,2)

120

In [90]:
%timeit binomial_coef_iter(160,20)

24 µs ± 465 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


We could improve the execution time from 62 to 25 micro seconds.

In [100]:
from pyccel.epyccel import epyccel
factorial_pyccel = epyccel(iter_factorial,language = 'c')
factorial_pyccel(2)


ERROR at shared library generation stage


RuntimeError: Failed to build module
In file included from /home/zineb/Documents/Al-Khwarizmi-HPC-Programming/day02/assignments/__epyccel__/__pyccel__/cwrapper/cwrapper.c:6:0:
/home/zineb/Documents/Al-Khwarizmi-HPC-Programming/day02/assignments/__epyccel__/__pyccel__/cwrapper/cwrapper.h:18:11: fatal error: Python.h: No such file or directory
 # include "Python.h"
           ^~~~~~~~~~
compilation terminated.


In [101]:
#accelerating using pyccel
from pyccel.epyccel import epyccel
factorial_pyccel = epyccel(iter_factorial,language = 'c')
#Binomial coefficient
def binomial_coef_iter(n : int, k : int):
    num = factorial_pyccel(n)
    den = factorial_pyccel(k) * factorial_pyccel(n - k)
    return num // den
binomial_coef_iter(16,2)
binom_coef = epyccel(binomial_coef_iter,language = 'c')
%timeit binom_coef(160,20)


ERROR at shared library generation stage


RuntimeError: Failed to build module
In file included from /home/zineb/Documents/Al-Khwarizmi-HPC-Programming/day02/assignments/__epyccel__/__pyccel__/cwrapper/cwrapper.c:6:0:
/home/zineb/Documents/Al-Khwarizmi-HPC-Programming/day02/assignments/__epyccel__/__pyccel__/cwrapper/cwrapper.h:18:11: fatal error: Python.h: No such file or directory
 # include "Python.h"
           ^~~~~~~~~~
compilation terminated.
